module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,206 @@
+)
+
+// Pruner is a component responsible for pruning data from
+// execution data storage. It is configured with the following
+// parameters:
+// * Height range target: The target number of most recent blocks
+//   to store data for. This controls the total amount of data
+//   stored on disk.
+// * Threshold: The number of block heights that we can exceed
+//   the height range target by before pruning is triggered. This
+//   controls the frequency of pruning.
+// The Pruner consumes a stream of fulfilled height notifications,
+// and triggers pruning once the difference between the fulfilled
+// height and the last pruned height reaches the height range
+// target + threshold.
+type Pruner struct {
+	storage       tracker.Storage
+	pruneCallback func(ctx context.Context) error
+
+	// channels used to send new fulfilled heights and config changes to the worker thread
+	// fulfilledHeights uses an unbounded channel to ensure NotifyFulfilledHeight is not blocking
+	fulfilledHeightsIn    chan<- interface{}
+	fulfilledHeightsOut   <-chan interface{}
+	thresholdChan         chan uint64
+	heightRangeTargetChan chan uint64
+
+	lastFulfilledHeight uint64
+	lastPrunedHeight    uint64
+
+	// the height range is the range of heights between the last pruned and last fulfilled
+	// heightRangeTarget is the target minimum value for this range, so that after pruning
+	// the height range is equal to the target.
+	heightRangeTarget uint64
+
+	// threshold defines the maximum height range and how frequently pruning is performed.
+	// once the height range reaches `heightRangeTarget+threshold`, `threshold` many blocks
+	// are pruned
+	threshold uint64
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataPrunerMetrics
+
+	component.Component
+	cm *component.ComponentManager
+}
+
+type PrunerOption func(*Pruner)
+
+// WithHeightRangeTarget is used to configure the pruner with a custom
+// height range target.
+func WithHeightRangeTarget(heightRangeTarget uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.heightRangeTarget = heightRangeTarget
+	}
+}
+
+// WithThreshold is used to configure the pruner with a custom threshold.
+func WithThreshold(threshold uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.threshold = threshold
+	}
+}
+
+func WithPruneCallback(callback func(context.Context) error) PrunerOption {
+	return func(p *Pruner) {
+		p.pruneCallback = callback
+	}
+}
+
+// NewPruner creates a new Pruner.
+func NewPruner(logger zerolog.Logger, metrics module.ExecutionDataPrunerMetrics, storage tracker.Storage, opts ...PrunerOption) (*Pruner, error) {
+	lastPrunedHeight, err := storage.GetPrunedHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pruned height: %w", err)
+	}
+
+	fulfilledHeight, err := storage.GetFulfilledHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulfilled height: %w", err)
+	}
+
+	fulfilledHeightsIn, fulfilledHeightsOut := util.UnboundedChannel()
+	fulfilledHeightsIn <- fulfilledHeight
```

Comment:
I have the same reservations about using unbounded channels here that I mentioned in the earlier iteration of this PR. In general I think unbounded channels are an antipattern that we should try to avoid. 

For this case, I think we can instead use a monotonous counter, since we only want to retain the largest fulfilled height we are notified of. 
https://github.com/onflow/flow-go/blob/3eb8166c369f335c0315e42569ccf2af66d9222f/engine/consensus/sealing/counters/monotonous_counter.go#L9

We use this in the sealing engine, and something similar in the Active Pacemaker. 
https://github.com/onflow/flow-go/blob/29c6cdefecc181c7726bac9ebcad9ee9ef3eea67/consensus/hotstuff/voteaggregator/vote_aggregator.go#L33

Reply:
I just ended up using a buffered channel for now, because if I used the counter I would still need to add an additional Notifier so that the worker routine could be notified every time there's an update